### PR TITLE
rpc_ppx extension

### PR DESCRIPTION
**Pull request by GuerfelMohamed**

### CODE REVIEW

File: src/ppx/ppx_eliom_utils.ml [hunk 1]
```diff
@@ -461,8 +463,120 @@ end
...
 *)
-module Shared = struct
+module Rpc = struct
+  let function_name fun_name_pattern =
+    let loc = fun_name_pattern.ppat_loc in
+    match fun_name_pattern with
+    | [%pat? [%p? {ppat_desc= Ppat_var ident}]] ->
+      ident.txt
+    | _ ->
+      Location.raise_errorf ~loc "syntaxe error: function name expected"
```
Comment (jrochel) [inline src/ppx/ppx_eliom_utils.ml:473]: syntax

File: src/ppx/ppx_eliom_utils.ml [hunk 2]
```diff
@@ -515,8 +523,10 @@ module Rpc = struct
       ->
       Location.raise_errorf ~loc:p.ppat_loc
         "Unexpected argument, (argument:type) format expected"
-    | _ ->
-      List.rev arg_list
+    | e -> begin match arg_list with 
+            |[]-> Location.raise_errorf ~loc:e.pexp_loc "No argements"
```
Comment (facelesspanda) [inline src/ppx/ppx_eliom_utils.ml:527]:
```suggestion
            |[]-> Location.raise_errorf ~loc:e.pexp_loc "No arguments"
```

File: src/ppx/ppx_eliom_utils.ml [hunk 3]
```diff
@@ -464,22 +464,27 @@ end
    ]
 *)
 module Rpc = struct
-  (*type error= 
+  type error=
     |No_arguments
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml]: I know there is no ocamlformat on eliom but you should try to follow the formatting conventions of the rest of the file. 
(spacing around operators, bars and so on)

Comment (remyzorg) [reply src/ppx/ppx_eliom_utils.ml]: Ok you cleaned it in the next commit.

File: src/ppx/ppx_eliom_utils.ml [hunk 4]
```diff
@@ -495,101 +501,242 @@ module Rpc = struct
...
+        ([ [%stri
+             let [%p pattern] =
+               [%expr
+                 fun [%p _args] ->
+                   [%e apply_function_expr ~loc ident.txt _apply]]]
+        ] [@metaloc loc])
     | _ -> (
         ([stri] [@metaloc loc]) )
+
+  (*let rpc_server rpc_type stri =
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:667]: I don't think this commented code is really needed.
If it is in a previous commit you can just get rid of it.

File: src/ppx/ppx_eliom_utils.ml [hunk 5]
```diff
@@ -285,6 +285,12 @@ module Context = struct
     | "client" | "client.start"
     | "eliom.client" | "eliom.client.start" -> `Client
     | _ -> invalid_arg "Eliom ppx: Not a context"
+  
+  let rpc_of_string = function
+    | "c_rpc"-> `Connected_rpc
+    | "c_rpc_o" -> `Connected_rpc_o
+    | "c_wrapper"-> `Connected_wrapper
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:292]: I would say :  
- rpc : wrapper
- cprc : connected rpc
- crpc_opt : optionnally connected rpc

what do you think @jrochel @ilankri @facelesspanda @vouillon @balat ?

File: src/ppx/ppx_eliom_utils.ml [hunk 6]
```diff
@@ -784,20 +931,29 @@ module Make (Pass : Pass) = struct
...
+            (flatmap (Rpc.client_side rpc_type) strs) 
+        in 
+        let l''= if rpc_type !=`Connected_wrapper
+        then let c=`Server in 
+              flatmap (dispatch_str c mapper)
+            (flatmap (Rpc.server_side rpc_type) strs)
+        else [] 
         in
-        let list = List.flatten [l; l'] in
         (*let migrate =
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:950]: dont forget to amend you commit to remove this

File: src/ppx/ppx_eliom_utils.ml (hunk 5 above)
Comment (facelesspanda) [reply src/ppx/ppx_eliom_utils.ml:292]: makes sense to me

File: src/ppx/ppx_eliom_utils.ml [hunk 7]
```diff
@@ -322,7 +330,7 @@ module type Pass = sig
   val client_sig: signature_item -> signature_item list
   val server_sig: signature_item -> signature_item list
 
-  (** How to handle "[%client ...]" and "[%shared ...]" expr. *)
+  (** How to handle "[\%client ...]" and "[\%shared ...]" expr. *)
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:333]: what are those ?

File: src/ppx/ppx_eliom_utils.ml [hunk 8]
```diff
@@ -654,6 +931,31 @@ module Make (Pass : Pass) = struct
         let c = Context.of_string txt in
         let l = flatmap (dispatch_str c mapper) strs in
         maybe_reset_injected_idents c ; l
+      | Pstr_extension (({txt}, PStr strs), _) 
+        when is_annotation txt ["c_wrapper";"c_rpc";"c_rpc_o"] ->
+        let rpc_type=Context.rpc_of_string txt in
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:936]: spaces after = everywhere.

File: src/ppx/ppx_eliom_utils.ml [hunk 9]
```diff
@@ -461,8 +469,277 @@ end
...
+            (Exp.ident (Location.mkloc (Longident.parse ident.txt) loc))
+            list
+        in
+        ([[%stri let [%p pattern] = fun [%p _args] -> [%e apply]]][@metaloc loc])
+    | _ ->
+        print_error ~loc Fatal_error )
+  | _ ->
+      print_error ~loc Fatal_error*)
+
+  let client_side rpc_type stri =
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:699]: I would advise you to prefix all functions building ast parts by "generate_".
As functions are actions to be done, it is nice to use a action verb to express this. This is not an absolute rule and there can be many exceptions. Here when I read `client_side`, before diving into your code I tell myself "client side what ?".

File: src/ppx/ppx_eliom_utils.ml [hunk 10]
```diff
@@ -461,8 +469,277 @@ end
...
+    [%expr
+      let aux =
+        ~%(Eliom_client.server_function ~name:[%e rpc_name]
+             [%json: [%t _typ_of_args]]
+             (Os_session.Opt.connected_rpc (fun [%p id] [%p _pat_of_args] ->
+                  [%e apply])))
+      in
+      [%e aux_apply]]
 
+  let expression fun_name_pattern expr apply_rpc =
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:599]: Again, more a naming matter but what kind of expression are you generating here ?
Is this the expression of the server side function ? If so, be more explicit.

File: src/ppx/ppx_eliom_utils.ml [hunk 11]
```diff
@@ -461,8 +469,277 @@ end
...
+      args_list
+
+  let rec args_parser expr arg_list =
+    match expr with
+    | [%expr fun ([%p? pattern] : [%t? typ]) -> [%e? expr']] ->
+        args_parser expr' ((pattern, typ, Asttypes.Nolabel) :: arg_list)
+    | [%expr fun () -> [%e? expr']] ->
+        args_parser expr'
+          ((AC.punit (), AC.tconstr "unit" [], Asttypes.Nolabel) :: arg_list)
+    | { pexp_desc=
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:522]:
```ocaml
| { pexp_desc=
￼          Pexp_fun
￼            ( arg
￼            , None
￼            , {ppat_desc= Ppat_constraint (pattern, typ)}
￼            , expr' ) } 
  ->
￼        args_parser expr' ((pattern, typ, arg) :: arg_list)
```
I think you could share a bit more code between labelled and optionnal args by doing something like that. But not sure what are the other enumerated type cases.

File: src/ppx/ppx_eliom_utils.ml (hunk 5 above)
Comment (ilankri) [reply src/ppx/ppx_eliom_utils.ml:292]: I am OK too with the suggested names.

File: src/ppx/ppx_eliom_utils.ml [hunk 12]
```diff
@@ -461,8 +469,277 @@ end
      ]
    ]
 *)
-module Shared = struct
+module Rpc = struct
+  type error = No_arguments | Missing_argument_type | Fatal_error
+
+  let print_error ~loc (e : error) =
+    let string_of_error =
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:476]: usually `string_of_something` are functions translating `something` to string in OCaml standards. `error_str` or something silimilare would fit a bit more to what you have have here.

File: src/ppx/ppx_eliom_utils.ml (hunk 11 above)
Comment (remyzorg) [reply src/ppx/ppx_eliom_utils.ml:522]: It seems to be only `Nolabel` and already matched before, so I think what I proposed you should be working.

File: src/ppx/ppx_eliom_utils.ml [hunk 13]
```diff
@@ -461,8 +469,277 @@ end
...
+
+  let connected_rpc_o ~id ~loc rpc_name fun_name args_list _pat_of_args _typ_of_args aux_apply
+      =
+    let apply =
+      apply_function_expr ~loc fun_name
+        ( (Asttypes.Nolabel, pat_to_exp id)
+        :: List.map
+             (fun (pattern, _, label) -> (label, pat_to_exp pattern))
+             args_list )
+    in
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:589]: this code block seem to be the same as connected rpc. There is maybe a possible factorization.

File: src/ppx/ppx_eliom_utils.ml [hunk 14]
```diff
@@ -461,8 +469,277 @@ end
...
+      |> format_args
+    in
+    let rec expr_mapper expr =
+      let loc = expr.pexp_loc in
+      match expr with
+      | [%expr fun ([%p? pattern] : [%t? _]) -> [%e? expr']] ->
+          [%expr fun [%p pattern] -> [%e expr_mapper expr']]
+      | [%expr fun () -> [%e? expr']] ->
+          [%expr fun () -> [%e expr_mapper expr']]
+      | {pexp_desc= Pexp_fun (Asttypes.Labelled s, None, pattern, expr')} ->
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml:618]: again, Labelled and Option are doing the same thing :
you could also write :  
```ocaml
| {pexp_desc= Pexp_fun (((Asttypes.(Labelled _ | Optional _) as arg), None, pattern, expr')} ->
          Exp.fun_ arg None pattern (expr_mapper expr')

```

File: src/ppx/ppx_eliom_utils.ml [hunk 15]
```diff
@@ -888,7 +915,7 @@ module Make (Pass : Pass) = struct
           flatmap (dispatch_str c mapper)
             (flatmap (Rpc.generate_client_struct_item rpc_type) strs) 
         in 
-        let l''= if rpc_type !=`Connected_wrapper
+        let l''= if (rpc_type ==`Connected_rpc || rpc_type ==`Connected_rpc_o)
```
Comment (ilankri) [inline src/ppx/ppx_eliom_utils.ml:918]: `( = )` is the structural equality whereas `( == )` is the physical one.

```suggestion
        let l''= if (rpc_type =`Connected_rpc || rpc_type =`Connected_rpc_o)
```

File: src/ppx/ppx_eliom_utils.ml [hunk 16]
```diff
@@ -461,8 +470,251 @@ end
...
+            Exp.constant (Pconst_string (rpc_name fun_name_pattern, None))
+          in
+          let fun_name = function_name fun_name_pattern in
+          apply_rpc ~loc (rpc_name, fun_name, _args_list, _pat_of_args, _typ_of_args, _expr_of_args)
+      |(pattern,_,label)::args_list' -> Exp.fun_ label None pattern (expr_mapper args_list')
+    in
+    expr_mapper _args_list [@metaloc loc]
+
+  let generate_server_struct_item rpc_type stri =
+    (*let loc = stri.pstr_loc in*)
```
Comment (ilankri) [inline src/ppx/ppx_eliom_utils.ml:643]: To be removed?

File: src/ppx/ppx_eliom_utils.ml [hunk 17]
```diff
@@ -654,6 +906,30 @@ module Make (Pass : Pass) = struct
...
+              flatmap (dispatch_str c mapper)
+            (flatmap (Rpc.generate_server_struct_item rpc_type) strs)
+        else [] 
+        in
+        let list = List.flatten [l;l';l''] in
+        (*let migrate =
+          Versions.migrate (module OCaml_408) (module OCaml_current)
+        in
+        Format.eprintf "%s@." @@ Pprintast.string_of_structure 
+        @@ migrate.Versions.copy_structure list ;*)
```
Comment (ilankri) [inline src/ppx/ppx_eliom_utils.ml:930]: To be removed?

File: src/ppx/ppx_eliom_utils.ml [hunk 18]
```diff
@@ -879,7 +906,7 @@ module Make (Pass : Pass) = struct
         let l = flatmap (dispatch_str c mapper) strs in
         maybe_reset_injected_idents c ; l
       | Pstr_extension (({txt}, PStr strs), _) 
-        when is_annotation txt ["rpc"; "crpc" ;"crpc_opt"] ->
```
Comment (remyzorg) [inline src/ppx/ppx_eliom_utils.ml]: The eliom rpc is supposed to be the default. I would specialize all ocsigenstart extension naming and name eliom one's "rpc".

"cw_rpc" might a good pick for connected wrapper.

**Comment by balat** [comment]
The documentation is not up to date.
Who is going to write it?

**Pull request merged**